diff --git a/isa/Makefile b/isa/Makefile
index bf85e1f..1e68ca1 100644
--- a/isa/Makefile
+++ b/isa/Makefile
@@ -45,7 +45,7 @@ default: all
 
 RISCV_PREFIX ?= riscv$(XLEN)-unknown-elf-
 RISCV_GCC ?= $(RISCV_PREFIX)gcc
-RISCV_GCC_OPTS ?= -static -mcmodel=medany -fvisibility=hidden -nostdlib -nostartfiles
+RISCV_GCC_OPTS ?= -static -mcmodel=medany -fvisibility=hidden -nostdlib -nostartfiles -Oz
 RISCV_OBJDUMP ?= $(RISCV_PREFIX)objdump --disassemble-all --disassemble-zeroes --section=.text --section=.text.startup --section=.text.init --section=.data
 RISCV_SIM ?= spike
 
diff --git a/isa/rv64si/dirty.S b/isa/rv64si/dirty.S
index 15f3163..73c8adf 100644
--- a/isa/rv64si/dirty.S
+++ b/isa/rv64si/dirty.S
@@ -17,6 +17,8 @@
 RVTEST_RV64M
 RVTEST_CODE_BEGIN
 
+  li sp, DRAM_BASE + 0x20000000
+
   # Turn on VM
   li a0, (SATP_MODE & ~(SATP_MODE<<1)) * SATP_MODE_SV39
   la a1, page_table_1
diff --git a/isa/rv64ua/amoadd_w.S b/isa/rv64ua/amoadd_w.S
index d076d45..05e28b8 100644
--- a/isa/rv64ua/amoadd_w.S
+++ b/isa/rv64ua/amoadd_w.S
@@ -12,6 +12,7 @@
 
 RVTEST_RV64U
 RVTEST_CODE_BEGIN
+RVTEST_WANT_AMO_EMU
 
   TEST_CASE(2, a4, 0xffffffff80000000, \
     li a0, 0xffffffff80000000; \
diff --git a/isa/rv64ua/amoand_w.S b/isa/rv64ua/amoand_w.S
index 7fe3bd0..06fe442 100644
--- a/isa/rv64ua/amoand_w.S
+++ b/isa/rv64ua/amoand_w.S
@@ -12,6 +12,7 @@
 
 RVTEST_RV64U
 RVTEST_CODE_BEGIN
+RVTEST_WANT_AMO_EMU
 
   TEST_CASE(2, a4, 0xffffffff80000000, \
     li a0, 0xffffffff80000000; \
diff --git a/isa/rv64ua/amomax_w.S b/isa/rv64ua/amomax_w.S
index 2c42982..0eb6005 100644
--- a/isa/rv64ua/amomax_w.S
+++ b/isa/rv64ua/amomax_w.S
@@ -12,6 +12,7 @@
 
 RVTEST_RV64U
 RVTEST_CODE_BEGIN
+RVTEST_WANT_AMO_EMU
 
   TEST_CASE(2, a4, 0xffffffff80000000, \
     li a0, 0xffffffff80000000; \
diff --git a/isa/rv64ua/amomaxu_w.S b/isa/rv64ua/amomaxu_w.S
index 6eabcd2..8673c0a 100644
--- a/isa/rv64ua/amomaxu_w.S
+++ b/isa/rv64ua/amomaxu_w.S
@@ -12,6 +12,7 @@
 
 RVTEST_RV64U
 RVTEST_CODE_BEGIN
+RVTEST_WANT_AMO_EMU
 
   TEST_CASE(2, a4, 0xffffffff80000000, \
     li a0, 0xffffffff80000000; \
diff --git a/isa/rv64ua/amomin_w.S b/isa/rv64ua/amomin_w.S
index 754f64d..5f43909 100644
--- a/isa/rv64ua/amomin_w.S
+++ b/isa/rv64ua/amomin_w.S
@@ -12,6 +12,7 @@
 
 RVTEST_RV64U
 RVTEST_CODE_BEGIN
+RVTEST_WANT_AMO_EMU
 
   TEST_CASE(2, a4, 0xffffffff80000000, \
     li a0, 0xffffffff80000000; \
diff --git a/isa/rv64ua/amominu_w.S b/isa/rv64ua/amominu_w.S
index d04a650..925470c 100644
--- a/isa/rv64ua/amominu_w.S
+++ b/isa/rv64ua/amominu_w.S
@@ -12,6 +12,7 @@
 
 RVTEST_RV64U
 RVTEST_CODE_BEGIN
+RVTEST_WANT_AMO_EMU
 
   TEST_CASE(2, a4, 0xffffffff80000000, \
     li a0, 0xffffffff80000000; \
diff --git a/isa/rv64ua/amoor_w.S b/isa/rv64ua/amoor_w.S
index e64b8c2..3dee9cd 100644
--- a/isa/rv64ua/amoor_w.S
+++ b/isa/rv64ua/amoor_w.S
@@ -12,6 +12,7 @@
 
 RVTEST_RV64U
 RVTEST_CODE_BEGIN
+RVTEST_WANT_AMO_EMU
 
   TEST_CASE(2, a4, 0xffffffff80000000, \
     li a0, 0xffffffff80000000; \
diff --git a/isa/rv64ua/amoswap_w.S b/isa/rv64ua/amoswap_w.S
index c4276dc..f73cb80 100644
--- a/isa/rv64ua/amoswap_w.S
+++ b/isa/rv64ua/amoswap_w.S
@@ -12,6 +12,7 @@
 
 RVTEST_RV64U
 RVTEST_CODE_BEGIN
+RVTEST_WANT_AMO_EMU
 
   TEST_CASE(2, a4, 0xffffffff80000000, \
     li a0, 0xffffffff80000000; \
diff --git a/isa/rv64ua/amoxor_w.S b/isa/rv64ua/amoxor_w.S
index 1b6fc48..6098aae 100644
--- a/isa/rv64ua/amoxor_w.S
+++ b/isa/rv64ua/amoxor_w.S
@@ -12,6 +12,7 @@
 
 RVTEST_RV64U
 RVTEST_CODE_BEGIN
+RVTEST_WANT_AMO_EMU
 
   TEST_CASE(2, a4, 0xffffffff80000000, \
     li a0, 0xffffffff80000000; \
diff --git a/isa/rv64ua/lrsc.S b/isa/rv64ua/lrsc.S
index 39fb156..93e2e4c 100644
--- a/isa/rv64ua/lrsc.S
+++ b/isa/rv64ua/lrsc.S
@@ -12,6 +12,7 @@
 
 RVTEST_RV64U
 RVTEST_CODE_BEGIN
+RVTEST_WANT_AMO_EMU
 
 # get a unique core id
 la a0, coreid
